python-frontend/src/main/java/org/sonar/python/semantic/v2/typeshed/TypeShedDescriptorsProvider.java

```diff
@@ -0,0 +1,143 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic.v2.typeshed;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import javax.annotation.CheckForNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.plugins.python.api.ProjectPythonVersion;
+import org.sonar.python.index.ClassDescriptor;
+import org.sonar.python.index.Descriptor;
+import org.sonar.python.index.ModuleDescriptor;
+import org.sonar.python.types.protobuf.SymbolsProtos.ModuleSymbol;
+
+import static org.sonar.plugins.python.api.types.BuiltinTypes.NONE_TYPE;
+
+public class TypeShedDescriptorsProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(TypeShedDescriptorsProvider.class);
+
+  private static final String PROTOBUF_BASE_RESOURCE_PATH = "/org/sonar/python/types/";
+  private static final String PROTOBUF_CUSTOM_STUBS = PROTOBUF_BASE_RESOURCE_PATH + "custom_protobuf/";
+  private static final String PROTOBUF = PROTOBUF_BASE_RESOURCE_PATH + "stdlib_protobuf/";
+  private static final String PROTOBUF_THIRD_PARTY = PROTOBUF_BASE_RESOURCE_PATH + "third_party_protobuf/";
+  private static final String PROTOBUF_THIRD_PARTY_MYPY = PROTOBUF_BASE_RESOURCE_PATH + "third_party_protobuf_mypy/";
+  public static final String BUILTINS_FQN = "builtins";
+  // This is needed for some Python 2 modules whose name differ from their Python 3 counterpart by capitalization only.
+  private static final Map<String, String> MODULES_TO_DISAMBIGUATE = Map.of(
+    "ConfigParser", "2@ConfigParser",
+    "Queue", "2@Queue",
+    "SocketServer", "2@SocketServer"
+  );
+  private final ModuleSymbolToDescriptorConverter moduleConverter;
+
+  private Set<String> supportedPythonVersions;
+  private Map<String, Descriptor> builtins;
+  private final Set<String> projectBasePackages;
+  private final Map<String, Map<String, Descriptor>> cachedDescriptors;
+
+  public TypeShedDescriptorsProvider(Set<String> projectBasePackages) {
+    moduleConverter = new ModuleSymbolToDescriptorConverter();
+    cachedDescriptors = new HashMap<>();
+    this.projectBasePackages = projectBasePackages;
+  }
+
+  //================================================================================
+  // Public methods
+  //================================================================================
+
+  public Map<String, Descriptor> builtinSymbols() {
+    if (builtins == null) {
+      supportedPythonVersions();
```

Comment:
I think the name of the function should be changed if the return value is never used.